src/Npgsql/ConnectorPool.cs

```diff
@@ -187,23 +176,26 @@ internal async ValueTask<NpgsqlConnector> AllocateLong(NpgsqlConnection conn, Np
         {
             // No idle connector was found in the pool.
             // We now loop until one of three things happen:
-            // 1. The pool isn't at max capacity (Total < Max), so we can create a new physical connection.
-            // 2. The pool is at maximum capacity and there are no idle connectors (Busy == Max),
+            // 1. The pool isn't at max capacity (Open < Max), so we can create a new physical connection.
+            // 2. The pool is at maximum capacity and there are no idle connectors (Open - Idle == Max),
             // so we enqueue an open attempt into the waiting queue, so that the next release will unblock it.
-            // 3. An connector makes it into the idle list (race condition with another Release().
+            // 3. An connector makes it into the idle list (race condition with another Release()).
             while (true)
             {
                 NpgsqlConnector? connector;
-                var openCount = Volatile.Read(ref State.Open);
 
+                var openCount = Volatile.Read(ref State.Open);
                 if (openCount < _max)
                 {
-                    var newOpenCount = openCount + 1;
-                    // We're under the pool's max capacity, try to "allocate" a slot for a new physical connection.
-                    if (Interlocked.CompareExchange(ref State.Open, newOpenCount, openCount) != openCount)
+                    // We're under the pool's max capacity, "allocate" a slot for a new physical connection.
+                    // Don't spin for this https://github.com/dotnet/coreclr/pull/21437
+                    while (true)
                     {
-                        // Our attempt to increment the open counter failed, loop again and retry.
-                        continue;
+                        openCount = Volatile.Read(ref State.Open);
+                        if (Interlocked.CompareExchange(ref State.Open, openCount + 1, openCount) == openCount)
+                        {
+                            break;
+                        }
```

Comment:
You referenced the issue where @benaadams advised an optimization. Why not to repeat it here?

```csharp
while (true)
{
    var openCountPrev = Interlocked.CompareExchange(ref State.Open, openCount + 1, openCount);
    if (openCountPrev == openCount) break;

    openCount = openCountPrev;
}
```

Reply:
Ah nice catch. I hadn't seen that folded code comment at all, just the post from Stephen. I'll change that!

Reply:
Fixed